### ISSUE TITLE
helper method to convert nn.Modules to multi-head models

### ISSUE THREAD

**Issue by AntonioCarta** [ContinualAI/avalanche#537]
Right now we need to change the `forward`. We should add some method to make the conversion easier whenever possible.

**Comment by AlbinSou** [comment]
Hello Antonio,

I would like to work on this feature. However, I'm not sure of the specifications. 

Would you like a function that, from a nn.Module already implementing forward(x, task_id), returns a MultiTaskModule where this forward function either overrides forward or forward_single_task (depending on whether the forward method accepts multiple task_ids or not)

Or should this function also add some kind of multitask classifier to a nn.Module that does not have any ? Like from a module that has classical linear classification layer you would additionally like to replace it by a MultiHeadClassifier. In which case from a nn.module that implements forward(x), you need to create a MultiTaskModule that now accepts task labels in the forward method.

**Comment by AntonioCarta** [comment]
This is something that was requested during our weekly meetings, if I remember correctly. Right now, if you want to change a single head model into a multihead, you have to subclass and change the forward method to add the `task_labels` argument:

```
class SimpleCNN(nn.Module):

    def __init__(self, num_classes=10):
        super(SimpleCNN, self).__init__()

        self.features = nn.Sequential(
            nn.Conv2d(3, 32, kernel_size=3, stride=1, padding=1),
            nn.ReLU(inplace=True),
            nn.Conv2d(32, 32, kernel_size=3, padding=0),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=2, stride=2),
            nn.Dropout(p=0.25),
            nn.Conv2d(32, 64, kernel_size=3, padding=1),
            nn.ReLU(inplace=True),
            nn.Conv2d(64, 64, kernel_size=3, padding=0),
            nn.ReLU(inplace=True),
            nn.MaxPool2d(kernel_size=2, stride=2),
            nn.Dropout(p=0.25),
            nn.Conv2d(64, 64, kernel_size=1, padding=0),
            nn.ReLU(inplace=True),
            nn.AdaptiveMaxPool2d(1),
            nn.Dropout(p=0.25)
        )
        self.classifier = nn.Sequential(
            nn.Linear(64, num_classes)
        )

    def forward(self, x):
        x = self.features(x)
        x = x.view(x.size(0), -1)
        x = self.classifier(x)
        return x


class MTSimpleCNN(SimpleCNN, MultiTaskModule):

    def __init__(self):
        """
            Multi-task CNN with multi-head classifier.
        """
        super().__init__()
        self.classifier = MultiHeadClassifier(64)

    def forward(self, x, task_labels):
        x = self.features(x)
        x = x.squeeze()
        x = self.classifier(x, task_labels)
        return x
```

What we would really like to have is:
```
model = SimpleCNN()
model = as_multitask(model, 'classifier')
```
where you don't have to create a new specialized subclass but you only need to give the name of the final layer.

You can take this issue. However, I'm not sure if this is really possible right now, I opened the issue only because others explitly asked for it. Maybe with some weird hack.

**Comment by AlbinSou** [comment]
I think this can be done using a decorator, for instance with something like this:

```
import torch
import torch.nn as nn

from avalanche.models.dynamic_modules import MultiTaskModule, MultiHeadClassifier

class MultiTaskDecorator(MultiTaskModule):
    def __init__(self, model, classifier_name):
        super().__init__()
        self.model = model

        old_classifier = getattr(model, classifier_name)

        # Replace model classifier by multihead
        setattr(self, classifier_name, 
        MultiHeadClassifier(old_classifier.in_features, 
        old_classifier.out_features))

        # Replace old classifier by empty block
        setattr(self.model, classifier_name, nn.Sequential())

    def forward_single_task(self, x, task_labels):  
        out = self.model(x)
        return self.classifier(out.view(out.size(0), -1), task_labels=task_labels)

def as_multitask(model, classifier_name):
    return MultiTaskDecorator(model, classifier_name)

class MLP(nn.Module):
    def __init__(self):
        super().__init__()
        self.layer1 = nn.Linear(784, 64)
        self.classifier = nn.Linear(64, 10)

    def forward(self, x):
        return self.classifier(self.layer1(x))


if __name__ == "__main__":
     
    ### This Works ->
    
    module = MLP()
    module = as_multitask(module, 'classifier')

    assert isinstance(module, MultiTaskModule)
    assert isinstance(getattr(module, 'classifier'), MultiHeadClassifier)

    test_input = torch.ones(1, 784)
    output = module(test_input, task_labels=0)
    print(output.shape)

    ### This doesn't work ->

    layer_weight = module.layer1.weight

```

However, I guess we must additionally leave access to classical model attributes, this can be done throug overriding the __getattr__ and __setattr__ fields, which is possible but needs to be done with caution (since pytorch nn.Module already overrides them). I can try to do that if you think that's a nice way of solving this problem.

**Comment by AntonioCarta** [comment]
This is a good solution, it should work with any pytorch model (e.g. torchvision's models).

> However, I guess we must additionally leave access to classical model attributes, this can be done throug overriding the getattr and setattr fields, which is possible but needs to be done with caution (since pytorch nn.Module already overrides them). I can try to do that if you think that's a nice way of solving this problem.

yes, if you can do it without breaking `nn.Module`.
